vhost-device-spi/src/vhu_spi.rs

```diff
@@ -0,0 +1,1206 @@
+    HandleEventUnknown,
+    #[error("Invalid descriptor size, expected: {0}, found: {1}")]
+    UnexpectedDescriptorSize(usize, u32),
+    #[error("Descriptor not found")]
+    DescriptorNotFound,
+    #[error("Descriptor read failed")]
+    DescriptorReadFailed,
+    #[error("Descriptor write failed")]
+    DescriptorWriteFailed,
+    #[error("Failed to send notification")]
+    NotificationFailed,
+    #[error("Failed to create new EventFd")]
+    EventFdFailed,
+    #[error("No memory configured")]
+    NoMemoryConfigured,
+}
+
+impl convert::From<Error> for io::Error {
+    fn from(e: Error) -> Self {
+        io::Error::new(io::ErrorKind::Other, e)
+    }
+}
+
+/// The final status written by the device
+const VIRTIO_SPI_TRANS_OK: u8 = 0;
+const VIRTIO_SPI_PARAM_ERR: u8 = 1;
+const VIRTIO_SPI_TRANS_ERR: u8 = 2;
+
+#[derive(Copy, Clone, Default, Debug)]
+#[repr(C)]
+struct VirtioSpiTransferHead {
+    chip_select_id: u8,
+    bits_per_word: u8,
+    cs_change: u8,
+    tx_nbits: u8,
+    rx_nbits: u8,
+    reserved1: u8,
+    reserved2: u8,
+    reserved3: u8,
+    mode: Le32,
+    freq: Le32,
+    word_delay_ns: Le32,
+    cs_setup_ns: Le32,
+    cs_delay_hold_ns: Le32,
+    cs_change_delay_inactive_ns: Le32,
+}
+// SAFETY: The layout of the structure is fixed and can be initialized by
+// reading its content from byte array.
+unsafe impl ByteValued for VirtioSpiTransferHead {}
+
+#[derive(Copy, Clone, Default)]
+#[repr(C)]
+struct VirtioSpiTransferResult {
+    status: u8,
+}
+// SAFETY: The layout of the structure is fixed and can be initialized by
+// reading its content from byte array.
+unsafe impl ByteValued for VirtioSpiTransferResult {}
+
+/// Virtio SPI Configuration
+#[derive(Copy, Clone, Debug, Default, PartialEq)]
+#[repr(C)]
+pub(crate) struct VirtioSpiConfig {
+    pub(crate) cs_max_number: u8,
+    pub(crate) cs_change_supported: u8,
+    pub(crate) tx_nbits_supported: u8,
+    pub(crate) rx_nbits_supported: u8,
+    pub(crate) bits_per_word_mask: Le32,
+    pub(crate) mode_func_supported: Le32,
+    pub(crate) max_freq_hz: Le32,
+    pub(crate) max_word_delay_ns: Le32,
+    pub(crate) max_cs_setup_ns: Le32,
+    pub(crate) max_cs_hold_ns: Le32,
+    pub(crate) max_cs_inactive_ns: Le32,
+}
+
+// SAFETY: The layout of the structure is fixed and can be initialized by
+// reading its content from byte array.
+unsafe impl ByteValued for VirtioSpiConfig {}
+
+pub(crate) struct VhostUserSpiBackend<D: SpiDevice> {
+    spi_ctrl: Arc<SpiController<D>>,
+    event_idx: bool,
+    pub exit_event: EventFd,
+    mem: Option<GuestMemoryAtomic<GuestMemoryMmap>>,
+}
+
+type SpiDescriptorChain = DescriptorChain<GuestMemoryLoadGuard<GuestMemoryMmap<()>>>;
+
+impl<D: SpiDevice> VhostUserSpiBackend<D> {
+    pub fn new(spi_ctrl: Arc<SpiController<D>>) -> Result<Self> {
+        Ok(VhostUserSpiBackend {
+            spi_ctrl,
+            event_idx: false,
+            exit_event: EventFd::new(EFD_NONBLOCK).map_err(|_| Error::EventFdFailed)?,
```

Comment:
Perhaps also include the error here to `EventFdFailed`.

Reply:
Sorry, I don't get it, could you please explain?

Reply:
I meant the change you commented elsewhere, to add the error in the unit variant:

```rust
enum Error {
...
EventFdFailed(_),
..
}
```

Reply:
I checked that EventFd::new failure doesn't have argument

Reply:
Hey @HaixuCui I missed your reply and just read it, sorry for the delay!

I meant that `EventFd::new` returns an `io::Error` on failure, and maybe we can make it a field in the spi crate's `Error::EventFdFailed` ? So that it can be reported to the user.

Reply:
I would use `std::io::Error` as the argument of `Error::EventFdFailed`, as

```
pub(crate) enum Error {
    #[error("Failed to create new EventFd with error {0:?}")]
    EventFdFailed(std::io::Error),
}

pub fn new(spi_ctrl: Arc<SpiController<D>>) -> Result<Self> {
        Ok(VhostUserSpiBackend {
            spi_ctrl,
            event_idx: false,
            exit_event: EventFd::new(EFD_NONBLOCK).map_err(Error::EventFdFailed)?,
            mem: None,
        })
    }
```

 while the error log shows some trait such as `Clone`, `std::cmp::Eq`, do I need to covert `std::io::Error` to some other type?